### DISCUSSION TITLE
Rook unable to reach Grafana with in-cluster name

### DISCUSSION THREAD

**Discussion by reefland** [rook/rook#11586] [Q&A]
In trying to setup Rook-Ceph / Grafana I kept getting a dashboard not found message.

Given this service:
```
$ k describe svc grafana -n monitoring
Name:              grafana
Namespace:         monitoring
Labels:            app.kubernetes.io/instance=kube-prometheus-stack
                   app.kubernetes.io/managed-by=Helm
                   app.kubernetes.io/name=grafana
                   app.kubernetes.io/version=9.3.1
                   argocd.argoproj.io/instance=kube-prometheus-stack
                   helm.sh/chart=grafana-6.48.2
Annotations:       <none>
Selector:          app.kubernetes.io/instance=kube-prometheus-stack,app.kubernetes.io/name=grafana
Type:              ClusterIP
IP Family Policy:  SingleStack
IP Families:       IPv4
IP:                10.43.174.7
IPs:               10.43.174.7
Port:              http-web  80/TCP
TargetPort:        3000/TCP
Endpoints:         10.42.0.6:3000
Session Affinity:  None
Events:            <none>
```

I tried to curl it from Rook toolbox, and I can't:
```
[root@k3s05 /]# curl http://grafana.monitoring
curl: (6) Could not resolve host: grafana.monitoring

[root@k3s05 /]# curl http://grafana.monitoring.svc.cluster.local
curl: (6) Could not resolve host: grafana.monitoring.svc.cluster.local
```

But can curl the service IP:
```
[root@k3s05 /]# curl http://10.43.174.7       
<a href="/grafana/login">Found</a>.
```

If I try to curl from a non-Rook-Ceph pod, works fine:
```
$ kubectl run curl --image=radial/busyboxplus:curl --rm=true --stdin=true --tty=true --restart=Never

If you don't see a command prompt, try pressing enter.

[ root@curl:/ ]$ curl http://grafana.monitoring.svc.cluster.local
<a href="/grafana/login">Found</a>.
```

I don't know what to do with this. Not sure why I have a DNS issue within Rook-Ceph pods to in-cluster services, but not in other pods.

In the mean time, I tried to set the External URL for Grafana:
```
ceph dashboard set-grafana-api-url https://k3s.[REDACTED].com/grafana/
```

It "works" in that I get dashboards, but also get error messages with it:
![image](https://user-images.githubusercontent.com/16230914/214740351-d3e15ec8-c378-48f3-b351-4320563adeb4.png)

Closeup:
![image](https://user-images.githubusercontent.com/16230914/214740943-261cbf1d-56c9-472d-bd4c-483b807c55b4.png)

Suggestions?

**Comment by travisn** [comment]
@galexrt Have you seen this issue in your grafana testing?

**Reply by galexrt** [reply]
@reefland Do you use host network for your CephCluster? Check the [DNS policy](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-dns-policy) on the Rook Ceph MGR Pods and/ or post the YAML here.

Additionally the non-Rook pod that you checked the dns resolution in, is it running on the same node as the Rook Ceph? Maybe it is a node related issue.
Also check if `cluster.local` is the actual "dns name" of your Kubernetes cluster, some clusters have it changed from the default.

**Comment by reefland** [comment]
@galexrt - Yes, I'm using Host Network for the CephCluster, to allow devices external to cluster to mount volumes.
```
            network:
              provider: host
```

I noticed that the Toolbox Pod already has:
```
    spec:
      dnsPolicy: ClusterFirstWithHostNet
```

And this is where I did my CURL testing from as documented above, still on the same node `k3s05`:
```
$ kubectl -n rook-ceph exec -it $(kubectl -n rook-ceph get pod -l "app=rook-ceph-tools" -o jsonpath='{.items[*].metadata.name}') -- bash

[root@k3s05 /]# curl http://grafana.monitoring.svc.cluster.local
curl: (6) Could not resolve host: grafana.monitoring.svc.cluster.local
```

Force a pod to run curl on same node:
```
$ kubectl run curl --image=radial/busyboxplus:curl --rm=true --stdin=true --tty=true --restart=Never \
  --overrides='{"apiVersion": "v1", "spec": {"nodeSelector": { "kubernetes.io/hostname": "k3s05" }}}'
     
[ root@curl:/ ]$ curl http://grafana.monitoring.svc.cluster.local
<a href="/grafana/login">Found</a>.
```

And it works. Being that works should mean `cluster.local` is the cluster name.

And to confirm both are running on same node:
```
$ kubectl get pods --all-namespaces -o wide --field-selector spec.nodeName=k3s05 | grep -E 'tools|curl'

kube-system      curl                                                1/1     Running     0              6m8s    10.42.4.72       k3s05   <none>           <none>
rook-ceph        rook-ceph-tools-794bddc7c8-ff4cx                    1/1     Running     0              10d     192.168.10.219   k3s05   <none>           <none>
```

**Reply by galexrt** [reply]
Please run a test pod that has `hostNetwork: true` to check if that might be the culprit of the issue.

My suspicion is that your CNI plugin might not fully "work" with host network out of the box as you expect, e.g., if I remember correctly Cillium needs some config tweaks to work better/ well with `hostNetwork`/ "host reachable services".
Which CNI plugin are you using? So no worries regarding your CNI till we got that confirmed.

Also please check the contents of the `/etc/resolv.conf` in the toolbox pod, the `kubectl run` test pod (`hostNetwork: false`) created and an additional test pod that has `hostNetwork: true`.

**Comment by reefland** [comment]
** Updated for test pod to include `dnsPolicy: ClusterFirstWithHostNet`**

I'm using the default Flannel used with K3s:
```
$ sudo  cat /var/lib/rancher/k3s/agent/etc/cni/net.d/10-flannel.conflist

{
  "name":"cbr0",
  "cniVersion":"1.0.0",
  "plugins":[
    {
      "type":"flannel",
      "delegate":{
        "hairpinMode":true,
        "forceAddress":true,
        "isDefaultGateway":true
      }
    },
    {
      "type":"portmap",
      "capabilities":{
        "portMappings":true
      }
    }
  ]
}
```

I test the curl pod using `hostNetwork: true` and `dnsPolicy: ClusterFirstWithHostNet` and it now fails:

```
$ kubectl run curl --image=radial/busyboxplus:curl --rm=true --stdin=true --tty=true --restart=Never \
  --overrides='{"apiVersion": "v1", "spec": {"hostNetwork": true, "dnsPolicy": "ClusterFirstWithHostNet"}}'

[ root@k3s05:/ ]$ curl http://grafana.monitoring.svc.cluster.local
curl: (6) Couldn't resolve host 'grafana.monitoring.svc.cluster.local'
```

While DNS fails, I can still curl the service directly and pod directly:
```
[ root@k3s05:/ ]$ curl http://10.43.174.7
<a href="/grafana/login">Found</a>.

[ root@k3s05:/ ]$ curl http://10.42.2.14:3000
<a href="/grafana/login">Found</a>.
```
---

Contents of Toolbox  `/etc/resolv.conf`:
```
[root@k3s05 /]# cat /etc/resolv.conf
search rook-ceph.svc.cluster.local svc.cluster.local cluster.local [REDACTED] localdomain
nameserver 10.43.0.10
options ndots:5

[root@k3s05 /]# curl http://grafana.monitoring.svc.cluster.local
curl: (6) Could not resolve host: grafana.monitoring.svc.cluster.local
```

* The `[REDACTED]` is my external domain name. 

Contents of CURL Test POD with just `hostNetwork: true`:
```
[ root@k3s05:/ ]$  cat /etc/resolv.conf
search [REDACTED] localdomain
nameserver 192.168.10.1

[ root@k3s05:/ ]$ curl http://grafana.monitoring.svc.cluster.local
curl: (6) Couldn't resolve host 'grafana.monitoring.svc.cluster.local'
```

Contents of CURL Test POD with `hostNetwork: true` and `dnsPolicy: ClusterFirstWithHostNet`:
```
[ root@k3s05:/ ]$ cat /etc/resolv.conf
search kube-system.svc.cluster.local svc.cluster.local cluster.local [REDACTED] localdomain
nameserver 10.43.0.10
options ndots:5

[ root@k3s05:/ ]$ curl http://grafana.monitoring.svc.cluster.local
curl: (6) Couldn't resolve host 'grafana.monitoring.svc.cluster.local'
```

Contents of CURL Test POD with `hostNetwork: false`:
```
[ root@curl:/ ]$ cat /etc/resolv.conf
search kube-system.svc.cluster.local svc.cluster.local cluster.local [REDACTED] localdomain
nameserver 10.43.0.10
options ndots:5

[ root@curl:/ ]$ curl http://grafana.monitoring.svc.cluster.local
<a href="/grafana/login">Found</a>.
```
---

I also have a single node cluster under vmware to test with, I can reproduce my issues in a simpler environment:
```
$ k get nodes
NAME        STATUS   ROLES                  AGE   VERSION
testlinux   Ready    control-plane,master   52d   v1.25.4+k3s1
```

Baseline test, single node test cluster:
```
$ kubectl run curl --image=radial/busyboxplus:curl --rm=true --stdin=true --tty=true --restart=Never

[ root@curl:/ ]$ curl http://grafana.monitoring.svc.cluster.local
<a href="/grafana/login">Found</a>.
```

Fails with hostNetwork: true
```
$ kubectl run curl --image=radial/busyboxplus:curl --rm=true --stdin=true --tty=true --restart=Never \
  --overrides='{"apiVersion": "v1", "spec": {"hostNetwork": true }}'

[ root@testlinux:/ ]$ curl http://grafana.monitoring.svc.cluster.local
curl: (6) Couldn't resolve host 'grafana.monitoring.svc.cluster.local'
```

**Reply by galexrt** [reply]
That is interesting as in my toolbox that has `hostNetwork: true` and `dnsPolicy: ClusterFirstWithHostNet`, the `resolv.conf` is using the internal cluster DNS server and not the node's `resolv.conf`.

My private clusters uses the local dns cache "addon" and the CNI is Cilium.

Not sure why your `resolv.conf` is different like that.

What Kubernetes version are you running? Maybe it is Kubernetes version related.

**Comment by reefland** [comment]
```
$ k get nodes
NAME    STATUS   ROLES                       AGE    VERSION
k3s01   Ready    control-plane,etcd,master   309d   v1.25.4+k3s1
k3s02   Ready    control-plane,etcd,master   282d   v1.25.4+k3s1
k3s03   Ready    control-plane,etcd,master   289d   v1.25.4+k3s1
k3s04   Ready    worker                      55d    v1.25.4+k3s1
k3s05   Ready    worker                      55d    v1.25.4+k3s1
```

Also posted in K3s:  <https://github.com/k3s-io/k3s/issues/6880>

**Reply by galexrt** [reply]
I see. Let's see what the k3s team is going to say. Please keep us updated, thanks!

**Reply by reefland** [reply]
I just upgraded to latest available on 1.25, and have better results:
```
$ k get nodes                                                                                                                                              127 ↵
NAME    STATUS   ROLES                       AGE    VERSION
k3s01   Ready    control-plane,etcd,master   312d   v1.25.6+k3s1
k3s02   Ready    control-plane,etcd,master   286d   v1.25.6+k3s1
k3s03   Ready    control-plane,etcd,master   293d   v1.25.6+k3s1
k3s04   Ready    worker                      58d    v1.25.6+k3s1
k3s05   Ready    worker                      58d    v1.25.6+k3s1
```

I can now curl internal service:
```
$ kubectl run curl --image=radial/busyboxplus:curl --rm=true --stdin=true --tty=true --restart=Never \
  --overrides='{"apiVersion": "v1", "spec": {"hostNetwork": true, "dnsPolicy": "ClusterFirstWithHostNet"}}'
  
[ root@k3s04:/ ]$ curl http://grafana.monitoring.svc.cluster.local
<a href="/grafana/login">Found</a>.
```

No change to '/etc/resolv.conf':
```
$ cat /etc/resolv.conf
search kube-system.svc.cluster.local svc.cluster.local cluster.local [REDACTED] localdomain
nameserver 10.43.0.10
options ndots:5
```

Now when I look at the dashboard, I briefly (less than 1 second maybe) get a rendered dashboard which looks full of data and it goes away with this:
![image](https://user-images.githubusercontent.com/16230914/217128003-5e53a1cc-e4d0-42da-9e22-c7c343333cb6.png)

I'll work on changing Rook config to use internal Grafana service name again, see if this issue goes away for me.

**Reply by galexrt** [reply]
@reefland Let us know how it goes when you use the internal Grafana service name now.

**Comment by reefland** [comment]
Attempt 1:
```
[root@k3s03 /]# ceph dashboard set-grafana-api-url 'http://grafana.monitoring.svc.cluster.local'
Option GRAFANA_API_URL updated

[root@k3s03 /]# ceph mgr module disable dashboard
[root@k3s03 /]# ceph mgr module enable dashboard
```
Result: `Grafana Dashboard doesn't exist.`

---

Attempt 2:
```
[root@k3s03 /]# ceph dashboard set-grafana-api-url 'http://grafana.monitoring.svc.cluster.local/grafana/'
Option GRAFANA_API_URL updated

[root@k3s03 /]# ceph mgr module disable dashboard
[root@k3s03 /]# ceph mgr module enable dashboard
```
Result:
![image](https://user-images.githubusercontent.com/16230914/217294509-0b89138d-dbab-4a69-8a5b-acd5b63e24b7.png)

---

Attempt 3:
```
[root@k3s03 /]# ceph dashboard get-grafana-frontend-api-url
[root@k3s03 /]# ceph dashboard set-grafana-frontend-api-url 'https://k3s.[REDACTED]/grafana/'
Option GRAFANA_FRONTEND_API_URL updated

[root@k3s03 /]# ceph mgr module disable dashboard
[root@k3s03 /]# ceph mgr module enable dashboard
```

Result:
![image](https://user-images.githubusercontent.com/16230914/217298232-7037935d-6a82-4c52-8027-9a6a5cb98137.png)

I have a dashboards... with error message.

![image](https://user-images.githubusercontent.com/16230914/217298785-a2686083-0354-486f-b3c9-6d3060b04f23.png)

--

I suspect I need to do something with `mgr/dashboard/standby_behaviour` perhaps? not sure what. 
```
[root@k3s03 /]# ceph config get mgr mgr/dashboard/standby_behaviour
redirect
```

**Reply by galexrt** [reply]
@reefland I don't know enough about the `standby_behaviour` but could you try setting it to `error` as described here https://docs.ceph.com/en/quincy/mgr/dashboard/#disable-the-redirection

@travisn Depending on the outcome of this, we might need to set it to `error` by default as described in the "Disable the redireciton" of the Ceph MGR dashboard docs link above.

**Reply by travisn** [reply]
Agreed, the standby behavior would be better as error, since the redirect doesn't work with K8s

**Reply by travisn** [reply]
Note that as seen in the doc [here](https://rook.io/docs/rook/latest/CRDs/Cluster/ceph-cluster-crd/#cluster-settings), services with `app=rook-ceph-mgr` will be updated by rook whenever the active mgr changes. 
```
Rook will automatically update all services (in the cluster namespace) that have a label app=rook-ceph-mgr with a selector 
pointing to the active mgr. This commonly applies to services for the dashboard or the prometheus metrics collector.
```

**Comment by reefland** [comment]
Didn't seem to make a difference:
```
[root@k3s03 /]# ceph config get mgr mgr/dashboard/standby_behaviour
redirect

[root@k3s03 /]# ceph config set mgr mgr/dashboard/standby_behaviour "error"
[root@k3s03 /]# ceph config set mgr mgr/dashboard/standby_error_status_code 503

[root@k3s03 /]# ceph mgr module disable dashboard
[root@k3s03 /]# ceph mgr module enable dashboard

[root@k3s03 /]# ceph config get mgr mgr/dashboard/standby_behaviour
error
```

Result:
![image](https://user-images.githubusercontent.com/16230914/217333797-3037f874-1e31-4993-a584-d8769c6a89b2.png)
